### PR TITLE
avocado.plugins: Add plugin to run job in docker container [v2]

### PR DESCRIPTION
**Pull request by ldoktor**
Similarly to --remote or --vm plugins this plugin allows one to run the
job inside a docker container by specifying the docker image. It
executes a new container, then attaches it and uses it similarly as
--remote plugin uses remote machine.

To check it out you can use "ldoktor/fedora-avocado" image which is
available on (the default) hub.docker.com

Trello: https://trello.com/c/FOxIriF3/616-support-running-tests-in-a-docker-container-based-on-the-remote-plugin
v1: https://github.com/avocado-framework/avocado/pull/1310

Changes:

```yaml
v2: Added --docker-cmd argument
v2: Removed the --docker-hypervisor-uri (superseded by --docker-cmd)
v2: New commit to handle urls without files correctly
v2: Fixed `docker-no-copy`
v2: Move `remote_test_dir` to class definition
```

### CODE REVIEW

File: avocado/core/remote/runner.py [hunk 1]
```diff
@@ -61,6 +61,9 @@ def _copy_files(self):
             url = self.job.urls[i]
             if not os.path.exists(url):     # use test_dir path + py
                 url = os.path.join(data_dir.get_test_dir(), url)
+            if not os.path.exists(url):
+                raise exceptions.JobError("Unable to map test id '%s' to file"
```
Comment (clebergnu) [inline avocado/core/remote/runner.py]: Works great.

File: avocado/plugins/docker.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    def configure(self, parser):
+        run_subcommand_parser = parser.subcommands.choices.get('run', None)
+        if run_subcommand_parser is None:
+            return
+
+        msg = 'test execution inside docker container'
+        cmd_parser = run_subcommand_parser.add_argument_group(msg)
+        cmd_parser.add_argument("--docker", help="Name of the docker image to"
+                                "run tests on.", metavar="IMAGE")
+        cmd_parser.add_argument("--docker-cmd", default="docker",
```
Comment (clebergnu) [inline avocado/plugins/docker.py]: Works great with my `sudo` based setup.

File: avocado/plugins/docker.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        # We need to create the base dir, otherwise docker creates it as root
+        self.remote.makedir(self.remote_test_dir)
+        self.job.log.info("DOCKER     : Container id '%s'"
+                          % self.remote.get_cid())
+        self.job.args.remote_no_copy = self.job.args.docker_no_copy
+
+    def tear_down(self):
+        self.remote.close()
+        if not self.job.args.docker_no_cleanup:
+            process.system("docker rm %s" % self.remote.get_cid())
```
Comment (clebergnu) [inline avocado/plugins/docker.py]: This needs a tweak:

```diff
diff --git a/avocado/plugins/docker.py b/avocado/plugins/docker.py
index 06bbe17..df3d430 100644
--- a/avocado/plugins/docker.py
+++ b/avocado/plugins/docker.py
@@ -131,9 +131,11 @@ class DockerTestRunner(RemoteTestRunner):
         self.job.args.remote_no_copy = self.job.args.docker_no_copy
 
     def tear_down(self):
+        dkrcmd = self.job.args.docker_cmd
         self.remote.close()
         if not self.job.args.docker_no_cleanup:
-            process.system("docker rm %s" % self.remote.get_cid())
+            process.system("%s rm %s" % (dkrcmd,
+                                         self.remote.get_cid())
 
 
 class Docker(CLI):
```

**Comment by clebergnu** [comment]
@ldoktor there's one issue that I've found.  When this job is shortcurt (in this case because of 9dda96572a2dd42213026a9dbb4af53d71fc8707, but it also apply to other circunstances) it leaves the containers running:

```
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
[cleber@localhost avocado]$ ./scripts/avocado run boot --docker-cmd 'sudo /usr/bin/docker' --docker ldoktor/fedora-avocado 
DOCKER     : Container id '3a6c60913745983d40c250eba8010f29ba196f6a921e016053146121b3bbcb63'

Avocado job failed: JobError: Unable to map test id 'boot' to file
[cleber@localhost avocado]$ docker ps
CONTAINER ID        IMAGE                    COMMAND             CREATED             STATUS              PORTS               NAMES
3a6c60913745        ldoktor/fedora-avocado   "bash"              4 seconds ago       Up 3 seconds                            nauseous_wescoff
[cleber@localhost avocado]$ ./scripts/avocado run boot --docker-cmd 'sudo /usr/bin/docker' --docker ldoktor/fedora-avocado 
DOCKER     : Container id '3da1f58968a59d600949f1f93c76375446188c9ecde52a7a749e7f7cc5b51f68'

Avocado job failed: JobError: Unable to map test id 'boot' to file
[cleber@localhost avocado]$ docker ps
CONTAINER ID        IMAGE                    COMMAND             CREATED             STATUS              PORTS               NAMES
3da1f58968a5        ldoktor/fedora-avocado   "bash"              3 seconds ago       Up 2 seconds                            loving_noether
3a6c60913745        ldoktor/fedora-avocado   "bash"              16 seconds ago      Up 16 seconds                           nauseous_wescoff
```

Basically the container is being started, and on job errors, they're left running.  Other case is with interruptions:

```
./scripts/avocado run sleeptenmin.py --docker-cmd 'sudo /usr/bin/docker' --docker ldoktor/fedora-avocado 
DOCKER     : Container id '240a26838a18603365efaa1646df57299cf21f7a6500b87068bf1f7f6d46f82d'
^C
Avocado crashed: ShellProcessTerminatedError: Shell process terminated while waiting for command to complete: 'cd /avocado_remote_test_dir; avocado run --force-job-id 4e322cc272a49bf6c542a5cfe8464daef1877daa --json - --archive /avocado_remote_test_dir/home/cleber/src/avocado/avocado/examples/tests/sleeptenmin.py '    (status: None,    output: '')
Traceback (most recent call last):

  File "/home/cleber/src/avocado/avocado/avocado/core/job.py", line 515, in _run
    self.args.test_result_total)

  File "/home/cleber/src/avocado/avocado/avocado/core/remote/runner.py", line 236, in run_suite
    results = self.run_test(self.job.urls, timeout)

  File "/home/cleber/src/avocado/avocado/avocado/core/remote/runner.py", line 161, in run_test
    timeout=timeout)

  File "/home/cleber/src/avocado/avocado/avocado/plugins/docker.py", line 92, in run
    print_func=print_func)

  File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1136, in cmd_status_output
    o = self.cmd_output(cmd, timeout, internal_timeout, print_func, safe)

  File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1059, in cmd_output
    raise ShellProcessTerminatedError(cmd, e.status, o)

ShellProcessTerminatedError: Shell process terminated while waiting for command to complete: 'cd /avocado_remote_test_dir; avocado run --force-job-id 4e322cc272a49bf6c542a5cfe8464daef1877daa --json - --archive /avocado_remote_test_dir/home/cleber/src/avocado/avocado/examples/tests/sleeptenmin.py '    (status: None,    output: '')

Please include the traceback info and command line used on your bug report
Report bugs visiting https://github.com/avocado-framework/avocado/issues/new
[cleber@localhost avocado]$ docker ps
CONTAINER ID        IMAGE                    COMMAND             CREATED             STATUS              PORTS               NAMES
240a26838a18        ldoktor/fedora-avocado   "bash"              9 seconds ago       Up 8 seconds                            determined_heyrovsky
3da1f58968a5        ldoktor/fedora-avocado   "bash"              3 minutes ago       Up 3 minutes                            loving_noether
3a6c60913745        ldoktor/fedora-avocado   "bash"              3 minutes ago       Up 3 minutes                            nauseous_wescoff
```

This may also be the case with the VM plugin, but there's are two subtle differences: 
1. with docker, the 10th failed command will have triggered 10 new containers, whereas with VM, it will check if one (and only one) VM is running (and if it's not, will start it).
2. with VM, the libvirt domain is intended to be kept running, with docker it's only intended to be kept running when `--docker-no-cleanup` is given.

This seems to actually be a limitation/bug of our "Test dispatcher" code, that fails to call a cleanup method on job errors.  So, I'm fine with merging this, if we document and plan acting on this limitation that is out of the scope of the plugin itself.

File: avocado/plugins/docker.py (hunk 3 above)
Comment (ldoktor) [inline avocado/plugins/docker.py]: Oh, I forgot that one (grep for docker was not really simple :smile:)

**Pull request closed without merging**